### ISSUE TITLE
Backgrounding with Ctrl-Z, then resuming, leaves gemacs in an unusable state

### ISSUE THREAD

**Issue by sgbeal** [glycerine/gemacs#4]
Try the following on a Unix-esque system (Linux, in my case):

1) Start up gemacs
2) Tap Ctrl-Z to background it for a moment while you go find a piece of information you need in the editor.
3) After finding that bit of info, type "fg" to foreground it.

Rather than restore the screen mode, it sets waiting for input on stdin. Ctrl-C kills it.

i suspect, without in any way being sure, that this is simply a matter of restoring the screen mode after resuming. How that's handled in Go, i haven't a clue (i'm a C guy).

My use case: my web hoster recently removed emacs from their shared servers and refuse to reinstall it, telling me i "will simply have to use vim." And yet 25 years of emacs muscle memory *Does Not Simply Use Vim*. So... gemacs is the only light-enough-weight alternative i've found for installing there. It often happens that i'm editing a file, need to background the editor to copy a filename or some such, and resume the session to paste that bit of info into the editor. That becomes impossible without running a second SSH session.

**Comment by glycerine** [comment]
@gdamore What's the recommended approach here-- should tcell clients be catching ctrl-Z / SIGTSTP and noticing when we come back... or something else?

**Comment by glycerine** [comment]
@sgbeal Stephan, thanks for the report. It is unfortunate. I messed with it for a little bit and can see that SIGCONT can be caught to notice resumption. However I had a little but not enough luck with re-initializing the terminal.  So this will have to stay open until someone more sophisticated about terminal semantics has an idea.

As a workaround, I suggest using the GNU screen or tmux terminal multiplexers and just leaving gemacs going. Start a 2nd terminal with screen and use that to get the side bits of info.

**Comment by sgbeal** [comment]
Thank you for looking into this! If a solution presents itself i'll be *thrilled*.

Unimportant sidebar: "The problem" with screen is that its default configuration interferes with (==intercepts, for its own purposes) emacs-style key combinations (which also means bash and 25-ish years of my well-honed finger-muscle memory) ;). My hoster apparently doesn't have screen installed, in any case, so even reconfiguring its braindead defaults isn't an option :/. Fortunately, though, this is not an everyday problem, and now that i know about it, my biggest enemy is not the ctrl-z behaviour itself, but my fingers reflexively wanting to tap ctrl-z before they tap ctrl-x ctrl-s ;).

**Comment by glycerine** [comment]
yes, the screen defaults must be rebound. Here is my usual .screenrc which remaps the main control key to Ctrl-z.  Two birds with one stone.
https://github.com/glycerine/screenrc

**Comment by sgbeal** [comment]
Thank you!

That leads to an interesting gemacs misbehaviour: i'm trying to copy/paste the URL to that repo into the screenrc so that i know where i got it later, and pasting `https://github.com/glycerine/screenrc` into gemacs stops pasting after `https://git`. 

To reproduce:

1) open gemacs
2) Select this URL *with your mouse*: `https://github.com/glycerine/screenrc` (optionally tap ctrl-c, but it's not necessary)
3) Try to paste it into gemacs by middle-clicking the mouse.

Result: `https://git` gets pasted.

*Weird.*

It's not specific to a URL: gemacs stops pasting after 11 characters (or bytes, i'm not certain which), regardless of what i paste that way: 012345678901234. Yup. 11 chars is the paste limit. i've reproduced this with two different terminal emulators (MATE terminal and xterm).

*Weird.*

**Comment by glycerine** [comment]
Yeah. Sorry. Not going to have a fix for that short-paste. Probably ever.

Maybe try configuring vim with emacs keybindings...

https://github.com/maxbrunsfeld/vim-emacs-bindings

https://unix.stackexchange.com/questions/49203/vim-customized-with-emacs-commands-insert-mode-only

**Comment by sgbeal** [comment]
Out of curiosity, what's the cause of the short paste? That seems like a weirdly specific limit.

**Comment by glycerine** [comment]
I don't know. You'd have to do the research.  Pseudo terminals can have line disciplines that can have archaic default limits would be my guess.
